### PR TITLE
Fixes #6003 by storing the signingKeyId instead of authorityKeyId

### LINKED ISSUE #6003: Leader stores incorrect signingKeyId in agent cache
#### Overview of the Issue

In some circumstances, Connect will reissue leaf certificates over and over again, even though they have not expired.

#### Reproduction Steps

1. Start Consul cluster with the default Consul CA.
2. Switch to Vault CA.
3. Observe a high number of certificates being issued.

#### Investigation

Debugging the cache logic shows that new leaf certs are being force-expired immediately. The logic in [`activeRootHasKey()`](https://github.com/hashicorp/consul/blob/a44541a91fdfdc12ee6c1fafe95aa38841d24a3c/agent/cache-types/connect_ca_leaf.go#L450-L462) does not find a match between the new leaf cert's signing key and any of the existing root certs.

This is due to logic in [`parseCAroot()`](https://github.com/hashicorp/consul/blob/a44541a91fdfdc12ee6c1fafe95aa38841d24a3c/agent/consul/leader.go#L1087) that is storing the root cert's authority key instead of its signing key.

For most CA root certs, the authority key and signing key would be the same, and storing `AuthorityKeyId` is fine. But when switching CA providers after leader election, Consul asks the new CA for an intermediate cert which is cross-signed by the outgoing CA. For an intermediate cert, `AuthorityKeyId` and `SigningKeyId` are not the same, leading the cache to believe the root cert has been rotated when it hasn't.

#### Log Messages

Adding some logging to `activeRootHasKey()` shows the problem:

```
2019-06-20T22:02:49.034-0400 [DEBUG] ca-leaf-cache: checking active roots for signing key 65:36:3a:61:62:3a:38:38:3a:66:30:3a:61:61:3a:35:30:3a:64:37:3a:66:62:3a:62:35:3a:66:65:3a:63:33:3a:62:36:3a:39:34:3a:63:63:3a:66:62:3a:62:35:3a:61:35:3a:39:62:3a:35:35:3a:38:34:3a:31:35:3a:30:31:3a:65:39:3a:38:31:3a:63:37:3a:62:30:3a:37:66:3a:66:32:3a:66:30:3a:30:61:3a:62:64:3a:65:34
2019-06-20T22:02:49.034-0400 [DEBUG] ca-leaf-cache: checking CA: name=Aws CA Root Cert active=true signingKeyID=e5:0a:e6:37:fd:e5:21:19:03:08:07:90:01:d9:88:86:80:05:f6:fc
2019-06-20T22:02:49.034-0400 [DEBUG] ca-leaf-cache: parsed CA cert: name=Consul CA cd97408c-9d13-b232-2b5e-fc1999bf1d56 signingKeyID=65:36:3a:61:62:3a:38:38:3a:66:30:3a:61:61:3a:35:30:3a:64:37:3a:66:62:3a:62:35:3a:66:65:3a:63:33:3a:62:36:3a:39:34:3a:63:63:3a:66:62:3a:62:35:3a:61:35:3a:39:62:3a:35:35:3a:38:34:3a:31:35:3a:30:31:3a:65:39:3a:38:31:3a:63:37:3a:62:30:3a:37:66:3a:66:32:3a:66:30:3a:30:61:3a:62:64:3a:65:34 authorityKeyID=e5:0a:e6:37:fd:e5:21:19:03:08:07:90:01:d9:88:86:80:05:f6:fc
```

The signing key held by the cache is actually the _authority key_ of the CA cert, so the keys do not match.

#### Solution

The proposed solution is to fix `parseCAroot()` to store the signing key ID, not the authority key ID. A PR will be submitted.

### PR DESCRIPTION
**Pull request by tradel**
Fixes #6003 by storing the signingKeyId instead of authorityKeyId.

### CODE REVIEW

**Comment by banks** [comment]
The CI failures are legit:

```
agent/consul/leader.go:1087:50: rootCert.SigningKeyId undefined (type *x509.Certificate has no field or method SigningKeyId)
```

File: agent/consul/leader.go [hunk 1]
```diff
@@ -1084,7 +1084,7 @@ func parseCARoot(pemValue, provider, clusterID string) (*structs.CARoot, error)
 		ID:                  id,
 		Name:                fmt.Sprintf("%s CA Root Cert", strings.Title(provider)),
 		SerialNumber:        rootCert.SerialNumber.Uint64(),
-		SigningKeyID:        connect.HexString(rootCert.AuthorityKeyId),
+		SigningKeyID:        connect.HexString(rootCert.SigningKeyId),
```
Comment (banks) [inline agent/consul/leader.go]: @tradel did this compile for you?

`rootCert` is a `*x509.Certificate` which has no field called `SigningKeyId`. https://golang.org/pkg/crypto/x509/#Certificate

The CI build doesn't compile this.

I this the confusion here comes from the fact that in the client we are handling responses of type `structs.CARoot` which has this `SigningKeyID` field. But that should correspond to the `AuthorityKeyID` of the _leaf_ cert.

I.e. each leaf should have an `AuthorityKeyID` which is the ID of the certificate that signed it.

So I _think_ the bug here is that for a self-signed root, the signing authority (`AuthorityKeyID`) and the actual root cert's ID (`SubjectKeyID`) are the same. But for an intermediate, they are different and we really want the `SubjectKeyID` here.

To confuse matters. We are using the word "Root" here to mean "a Connect trust root" and something that can sign certs. It might not actually be an x509 root and so may be an intermediate cert. In that case we should be using the ID of the signing cert (the one in the `pemValue` here) as the `SigningKeyID` not the ID of they key that signed the "root".

tl;dr:
```suggestion
		SigningKeyID:        connect.HexString(rootCert.SubjectKeyID),
```

But tests would be good too.

**Comment by i0rek** [comment]
@banks I added a test to make sure it is doing what it should be. Is this what you've been asking for?
